File: DesignPatternsExercises/java/FlyweightDesignPattern/Solution1/OakTree.java
// public class OakTree __________(3)__________  Tree {
public class OakTree extends Tree {
    private int rootDepth;
    private String forestType;

    public OakTree() {
        // __________(4)__________ (false, false, 70);
        super(false, false, 70);
        forestType = "Deciduous";
    }

    @Override
    public void changeProperty() {
        System.out.println("Changing Oak Tree property Forest Type to Evergreen\n");
        setForestType("Evergreen");
    }

    public int getRootDepth() {
        return rootDepth;
    }

    public void setRootDepth(int rootDepth) {
        this.rootDepth = rootDepth;
    }

    public String getForestType() {
        return forestType;
    }

    public void setForestType(String forestType) {
        this.forestType = forestType;
    }

    @Override
    public String toString() {
        return "Oak Tree ~~ Height: " + getHeight() + "  Flowers: " + hasFlowers() + "  Fruits: " + hasFruits() +
                "  Root Depth: " + getRootDepth() + "  Forest Type: " + getForestType();
    }
}

File: DesignPatternsExercises/ruby/flyweight_design_pattern/flyweight_design_pattern1/exercise/oak_tree.rb
require_relative 'tree'

class OakTree < __________(5)__________
  attr_accessor :root_depth

  def __________(6)__________
    __________(7)__________(false, false, 70)
    @forest_type = :Deciduous
    @root_depth = 0
  end

  def __________(8)__________
    puts 'Changing Oak Tree property Forest Type to Evergreen'
    puts
    @forest_type = :Evergreen
  end

  def to_string
    "Oak Tree ~~ Height: #{@height}  Flowers: #{@has_flowers}  Fruits: #{@has_fruits}  Root Depth: #{@root_depth}  Forest Type: #{@forest_type}"
  end
end
